### ISSUE TITLE
GCP-VQVAE roundtrip reconstruction gives very high RMSD on both PDB and official CAMEO2024   HDF5 inputs

### ISSUE THREAD

**Issue by liukkcggvb** [mahdip72/vq_encoder_decoder#16]
Hi, thanks for releasing GCP-VQVAE.

I am trying to verify basic roundtrip reconstruction quality with the public code/checkpoints,
but I consistently get very poor backbone reconstruction. This seems related to #15, but I
wanted to provide a minimal reproduction with concrete numbers.

## Environment

I tested with the latest package installed from 'pip install "git+https://github.com/mahdip72/vq_encoder_decoder.git@master#subdirectory=gcp-vqvae"'.

```text
gcp-vqvae: 0.2.1
source: current master, subdirectory gcp-vqvae/gcp_vqvae
torch: 2.8.0+cu129
torch_cluster: 1.6.3+pt28cu129
torch_scatter: 2.1.2+pt28cu129
numpy: 1.26.4
CUDA available: true

pip check reports no broken requirements.

## Model

Local lite checkpoint/configs:

trained_model_dir=/path/to/gcp-vqvae-lite
checkpoint_path=checkpoints/best_valid.pth
config_vqvae=config_vqvae.yaml
config_encoder=config_gcpnet_encoder.yaml
config_decoder=config_geometric_decoder.yaml

The checkpoint loads cleanly:

missing keys: 0
unexpected keys: 0

I also previously tried the Hugging Face lite/large models and observed similarly poor
reconstruction.

## Test 1: PDB/CIF pdb_dir wrapper path

I first tested the standalone package path that reads PDB files directly through pdb_dir.

Input: a clean single-chain protein PDB.

sample: 7y3n antigen chain A
sequence length: 196
native CA-CA mean: 3.8046 Å
native CA-CA min: 2.9935 Å
native CA-CA max: 3.8927 Å
native CA-CA > 5 Å: 0

Minimal script:

import json
import torch
from gcp_vqvae import GCPVQVAE

def kabsch_align(mobile, target):
    mobile = torch.as_tensor(mobile, dtype=torch.float32)
    target = torch.as_tensor(target, dtype=torch.float32)
    mobile_center = mobile.mean(dim=0, keepdim=True)
    target_center = target.mean(dim=0, keepdim=True)
    mobile0 = mobile - mobile_center
    target0 = target - target_center
    cov = mobile0.T @ target0
    u, _, vh = torch.linalg.svd(cov)
    rot = vh.T @ u.T
    if torch.linalg.det(rot) < 0:
        vh = vh.clone()
        vh[-1, :] *= -1
        rot = vh.T @ u.T
    return mobile0 @ rot.T + target_center

def kabsch_rmsd(pred, true):
    pred = torch.as_tensor(pred, dtype=torch.float32).reshape(-1, 3)
    true = torch.as_tensor(true, dtype=torch.float32).reshape(-1, 3)
    aligned = kabsch_align(pred, true)
    return float(torch.sqrt(torch.mean(torch.sum((aligned - true) ** 2, dim=-1))).item())

def ca_stats(coords):
    coords = torch.as_tensor(coords, dtype=torch.float32)
    ca = coords[:, 1, :]
    distances = torch.linalg.norm(ca[1:] - ca[:-1], dim=-1)
    return {
        "mean": float(distances.mean().item()),
        "min": float(distances.min().item()),
        "max": float(distances.max().item()),
        "gt5": int((distances > 5.0).sum().item()),
        "gt10": int((distances > 10.0).sum().item()),
    }

pdb_dir = "/path/to/one_single_chain_pdb_dir"
trained_model_dir = "/path/to/gcp-vqvae-lite"

model = GCPVQVAE(
    trained_model_dir=trained_model_dir,
    use_hf=False,
    mode="all",
    mixed_precision="no",
    deterministic=True,
    suppress_logging=False,
)

dataset, collate_fn = model._build_dataset(
    pdb_dir=pdb_dir,
    max_task_samples=1,
    progress=True,
)
assert len(dataset) == 1

batch = collate_fn([dataset[0]])
for key in ("graph", "masks", "nan_masks", "target_coords"):
    batch[key] = batch[key].to(model.device)

with torch.inference_mode():
    full_output = model.model(batch)
    vq_output = model.model(batch, return_vq_layer=True)

seq_len = len(batch["seq"][0])
valid = (batch["masks"][0, :seq_len] & batch["nan_masks"][0, :seq_len]).detach().cpu()
true = batch["target_coords"].view(1, model.max_length, 3, 3)[0, :seq_len].detach().cpu()
full_pred = full_output["outputs"].view(1, model.max_length, 3, 3)[0, :seq_len].detach().cpu()
indices = vq_output["indices"][0, :seq_len].detach().cpu().tolist()

decode_output = model.decode([indices], pids=["decode_only"], batch_size=1)
decode_pred = decode_output["coords"][0][:seq_len].detach().cpu()

result = {
    "pid": str(batch["pid"][0]),
    "sequence_length": int(seq_len),
    "valid_length": int(valid.sum().item()),
    "invalid_indices": int(sum(1 for token in indices if int(token) < 0)),
    "full_model_rmsd_to_true": kabsch_rmsd(full_pred[valid], true[valid]),
    "decode_indices_rmsd_to_true": kabsch_rmsd(decode_pred[valid], true[valid]),
    "decode_indices_rmsd_to_full_model": kabsch_rmsd(decode_pred[valid], full_pred[valid]),
    "true_ca_stats": ca_stats(true[valid]),
    "full_model_ca_stats": ca_stats(full_pred[valid]),
    "decode_indices_ca_stats": ca_stats(decode_pred[valid]),
    "indices_min": int(min(indices)),
    "indices_max": int(max(indices)),
}

print(json.dumps(result, indent=2))

Actual output:

{
  "pid": "0_7y3n_antigen_A",
  "sequence_length": 196,
  "valid_length": 196,
  "invalid_indices": 0,
  "full_model_rmsd_to_true": 17.31204605102539,
  "decode_indices_rmsd_to_true": 18.533323287963867,
  "decode_indices_rmsd_to_full_model": 14.221552848815918,
  "true_ca_stats": {
    "mean": 3.8046305179595947,
    "min": 2.9934513568878174,
    "max": 3.8927080631256104,
    "gt5": 0,
    "gt10": 0
  },
  "full_model_ca_stats": {
    "mean": 7.389586925506592,
    "min": 0.995024561882019,
    "max": 42.2859001159668,
    "gt5": 125,
    "gt10": 37
  },
  "decode_indices_ca_stats": {
    "mean": 7.904085159301758,
    "min": 1.028112769126892,
    "max": 29.35640525817871,
    "gt5": 137,
    "gt10": 53
  },
  "indices_min": 4,
  "indices_max": 4072
}

## Test 2: official CAMEO2024 HDF5 path

I also tested the official CAMEO2024 HDF5 dataset path to rule out PDB preprocessing issues.

I downloaded cameo_2048_h5 and selected a clean sample:

sample: 0_8R7Y_A.h5
sequence length: 312
contains X: no
contains NaN backbone coords: no
native CA-CA mean: 3.8127 Å
native CA-CA min: 3.7644 Å
native CA-CA max: 3.9039 Å
native CA-CA > 5 Å: 0

This test uses the HDF5-style path used by the official evaluation.py:

GCPNetDataset -> custom_collate_pretrained_gcp -> model(batch)

The HDF5 file contains the expected fields:

seq
N_CA_C_O_coord
plddt_scores

Minimal HDF5 script:

import json
import torch
from gcp_vqvae import GCPVQVAE
from gcp_vqvae._internal.data.dataset import GCPNetDataset, custom_collate_pretrained_gcp

def kabsch_align(mobile, target):
    mobile = torch.as_tensor(mobile, dtype=torch.float32)
    target = torch.as_tensor(target, dtype=torch.float32)
    mobile_center = mobile.mean(dim=0, keepdim=True)
    target_center = target.mean(dim=0, keepdim=True)
    mobile0 = mobile - mobile_center
    target0 = target - target_center
    cov = mobile0.T @ target0
    u, _, vh = torch.linalg.svd(cov)
    rot = vh.T @ u.T
    if torch.linalg.det(rot) < 0:
        vh = vh.clone()
        vh[-1, :] *= -1
        rot = vh.T @ u.T
    return mobile0 @ rot.T + target_center

def kabsch_rmsd(pred, true):
    pred = torch.as_tensor(pred, dtype=torch.float32).reshape(-1, 3)
    true = torch.as_tensor(true, dtype=torch.float32).reshape(-1, 3)
    aligned = kabsch_align(pred, true)
    return float(torch.sqrt(torch.mean(torch.sum((aligned - true) ** 2, dim=-1))).item())

def ca_stats(coords):
    coords = torch.as_tensor(coords, dtype=torch.float32)
    ca = coords[:, 1, :]
    distances = torch.linalg.norm(ca[1:] - ca[:-1], dim=-1)
    return {
        "mean": float(distances.mean().item()),
        "min": float(distances.min().item()),
        "max": float(distances.max().item()),
        "gt5": int((distances > 5.0).sum().item()),
        "gt10": int((distances > 10.0).sum().item()),
    }

h5_dir = "/path/to/single_h5_dir_containing_0_8R7Y_A.h5"
trained_model_dir = "/path/to/gcp-vqvae-lite"

model = GCPVQVAE(
    trained_model_dir=trained_model_dir,
    use_hf=False,
    mode="all",
    mixed_precision="no",
    deterministic=True,
    suppress_logging=False,
)

dataset = GCPNetDataset(
    h5_dir,
    top_k=model.encoder_configs.top_k,
    num_positional_embeddings=model.encoder_configs.num_positional_embeddings,
    encoder_config_path=model.encoder_config_path,
    configs=model.configs,
    mode="evaluation",
)

collate_fn = lambda batch: custom_collate_pretrained_gcp(
    batch,
    featuriser=dataset.pretrained_featuriser,
    task_transform=dataset.pretrained_task_transform,
)

batch = collate_fn([dataset[0]])
for key in ("graph", "masks", "nan_masks", "target_coords"):
    batch[key] = batch[key].to(model.device)

with torch.inference_mode():
    full_output = model.model(batch)
    vq_output = model.model(batch, return_vq_layer=True)

seq_len = len(batch["seq"][0])
valid = (batch["masks"][0, :seq_len] & batch["nan_masks"][0, :seq_len]).detach().cpu()
true = batch["target_coords"].view(1, model.max_length, 3, 3)[0, :seq_len].detach().cpu()
full_pred = full_output["outputs"].view(1, model.max_length, 3, 3)[0, :seq_len].detach().cpu()
indices = vq_output["indices"][0, :seq_len].detach().cpu().tolist()

decode_output = model.decode([indices], pids=["decode_only"], batch_size=1)
decode_pred = decode_output["coords"][0][:seq_len].detach().cpu()

result = {
    "pid": str(batch["pid"][0]),
    "sequence_length": int(seq_len),
    "valid_length": int(valid.sum().item()),
    "invalid_indices": int(sum(1 for token in indices if int(token) < 0)),
    "full_model_rmsd_to_true": kabsch_rmsd(full_pred[valid], true[valid]),
    "decode_indices_rmsd_to_true": kabsch_rmsd(decode_pred[valid], true[valid]),
    "decode_indices_rmsd_to_full_model": kabsch_rmsd(decode_pred[valid], full_pred[valid]),
    "true_ca_stats": ca_stats(true[valid]),
    "full_model_ca_stats": ca_stats(full_pred[valid]),
    "decode_indices_ca_stats": ca_stats(decode_pred[valid]),
    "indices_min": int(min(indices)),
    "indices_max": int(max(indices)),
}

print(json.dumps(result, indent=2))

Actual output:

{
  "pid": "0_8R7Y_A",
  "sequence_length": 312,
  "valid_length": 312,
  "invalid_indices": 0,
  "full_model_rmsd_to_true": 23.161815643310547,
  "decode_indices_rmsd_to_true": 24.793363571166992,
  "decode_indices_rmsd_to_full_model": 14.42635726928711,
  "true_ca_stats": {
    "mean": 3.8127284049987793,
    "min": 3.7644290924072266,
    "max": 3.9039385318756104,
    "gt5": 0,
    "gt10": 0
  },
  "full_model_ca_stats": {
    "mean": 6.995386600494385,
    "min": 1.1461656093597412,
    "max": 20.658248901367188,
    "gt5": 209,
    "gt10": 58
  },
  "decode_indices_ca_stats": {
    "mean": 8.287474632263184,
    "min": 0.4182854890823364,
    "max": 58.49195861816406,
    "gt5": 219,
    "gt10": 83
  },
  "indices_min": 6,
  "indices_max": 4026
}

## Expected behavior

Based on the reported reconstruction RMSD in the README/paper, I expected the backbone RMSD to
be close to sub-angstrom or at least structurally reasonable.

Instead, both:

1. the PDB/CIF pdb_dir wrapper path, and
2. the official CAMEO2024 HDF5 path

produce disconnected backbones with very large RMSD and many adjacent CA distances larger than 5
Å.

Looking forward to hearing from you. Thanks!
